File: 3332.旅客可以得到的最多点数.cpp
/*
 * @lc app=leetcode.cn id=3332 lang=cpp
 *
 * [3332] 旅客可以得到的最多点数
 */

// @lc code=start
class Solution
{
public:
    int maxScore(int n, int k, vector<vector<int>> &stayScore, vector<vector<int>> &travelScore)
    {
        vector<vector<int>> dp(k + 1, vector<int>(n, 0));
        // 状态转移
        for (int i = 1; i <= k; i++)
            for (int j = 0; j < n; j++)
            {
                // 留在当前城市 j
                dp[i][j] = dp[i - 1][j] + stayScore[i - 1][j];
                // 来自另外一座城市 d
                for (int d = 0; d < n; d++)
                    dp[i][j] = max(dp[i][j], dp[i - 1][d] + travelScore[d][j]);
            }

        int ans = INT_MIN;
        // 枚举城市 i 作为终点
        for (int i = 0; i < n; i++)
            ans = max(ans, dp[k][i]);

        return ans;
    }
};
// @lc code=end

File: 3332.旅客可以得到的最多点数.py
#
# @lc app=leetcode.cn id=3332 lang=python3
#
# [3332] 旅客可以得到的最多点数
#

# @lc code=start
class Solution:
    def maxScore(self, n: int, k: int, stayScore: List[List[int]], travelScore: List[List[int]]) -> int:
        
        @cache
        def dfs(i: int, j: int) -> int:
            if i == k:
                return 0

            # 留在当前城市 j
            res1 = dfs(i + 1, j) + stayScore[i][j]
            # 前往另外一座城市 d
            res2 = max(dfs(i + 1, d) + s for d, s in enumerate(travelScore[j]))
            
            return max(res1, res2)
        
        return max(dfs(0, i) for i in range(n))  # 枚举城市 i 作为起点
# @lc code=end

